### DISCUSSION TITLE
Github Action caching/optimizations.

### DISCUSSION THREAD

**Discussion by SoffiMeower** [obhq/obliteration#40] [General]
I have it where QT auto caches, Vulkan SDK caches as we have it on a hardset version, and VCPKG keeps a cache until any dependency files are changed. If they're changed, VCPKG runs again and caches the new dependencies. Either way... Any other things we can cache for faster action times? I was not able to have Vulkan cache for linux as it's an Apt install.

Action times: These are builds re-executed to test with cache.
```
BUILD #    |             CACHES            | Win (With Build) | Win (No Build) 
Build # 47 |              NONE             | 7m 11s           | 3m 09s
Build # 54 |         QT + Vulkan           | 6m 46s           | 2m 34s (-35s) 
Build # 58 |      QT + Vulkan + VCPKG      | 6m 11s           | 1m 57s (-37s)
Build # 60 |  QT + Vulkan + VCPKG + CMAKE  | 5m 03s           | 1m 11s (-46s)
```

**Comment by ultimaweapon** [comment]
That should be the best possible we can get.

**Comment by SoffiMeower** [comment]
Unless CMake can only recompile changed files (Because we can store the whole Cmake cache and still have the command run.), then I'd agree.

**Reply by ultimaweapon** [reply]
Actually that is the default behavior, including Cargo. But I feel uncomfortable to cache this because if GitHub have updated toolchains on the runner and CMake decided to use the cached version instead of re-compile the whole project we are going to have strange bugs.

**Reply by SoffiMeower** [reply]
You can cache specific folders, and, even then, we can exclude other folders as well with !~/the/folderpath/here

**Reply by ultimaweapon** [reply]
What folders you are planing to cache?

**Reply by SoffiMeower** [reply]
That's up to you. I still can't build locally sadly... CMake says the CMakeList can't find the Qt folder.

**Reply by SoffiMeower** [reply]
Caching uses a regular windows and linux folder path, so if there was something you wanted excluded from Build, we'd do...
```
Keep:
D:\a\Obliteration\Build\
Exclude:
!D:\a\Obliteration\Build\dontuse\
```

**Reply by ultimaweapon** [reply]
There are 2 folders related to this, `build` and `src/target`; both of them must be re-compile when the toolchains have been updated.

For Qt try to set `Qt6_DIR` environment variable to the full path where Qt is installed.

**Reply by SoffiMeower** [reply]
> There are 2 folders related to this, `build` and `src/target`; both of them must be re-compile when the toolchains have been updated.
> 
> For Qt try to set `Qt6_DIR` environment variable to the full path where Qt is installed.

As you've been seeing, it checks its environment (runner OS) when making the Cache Key. So if the OS/Environment changes, so does the Key, which then invalidates the cache and makes it request a recache/rerun.

**Reply by SoffiMeower** [reply]
And, would I want this with both the Run Cmake step and Build step, or just the Build step?

**Reply by ultimaweapon** [reply]
I mean setting `Qt6_DIR` on your computer to fix CMake can't find Qt folder.

**Comment by SoffiMeower** [comment]
The cache can be up to 10GB. Anyways, any questions or suggestions?

**Comment by SoffiMeower** [comment]
@ultimaweapon Chart updated with Cmake cache added! More than a full minute decrease from the last build time!

**Comment by SoffiMeower** [comment]
New change. Note: Changing the workflow YMLs or Dependency files will invalidate the Cmake cache due to the build environment being changed.

**Reply by ultimaweapon** [reply]
It is also possible for GitHub to update pre-installed software on the runner and it is possible for the software to be a different version when we install via `apt-get`. Whenever this happen we must make sure the whole project get re-compiled, not using the cached data.
